x/migration/keeper/msg_server_claim_morse_supplier.go

```diff
@@ -2,17 +2,146 @@ package keeper
+	// from the claimed balance.
+	if err := k.MintClaimedMorseTokens(ctx, shannonAccAddr, morseClaimableAccount.TotalTokens()); err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	// Set ShannonDestAddress & ClaimedAtHeight (claim).
+	morseClaimableAccount.ShannonDestAddress = shannonAccAddr.String()
+	morseClaimableAccount.ClaimedAtHeight = sdkCtx.BlockHeight()
+
+	// Update the MorseClaimableAccount.
+	k.SetMorseClaimableAccount(
+		sdkCtx,
+		morseClaimableAccount,
+	)
+
+	msgStakeSupplier := suppliertypes.NewMsgStakeSupplier(
+		shannonAccAddr.String(),
+		shannonAccAddr.String(),
+		shannonAccAddr.String(),
+		*msg.Stake,
+		[]*sharedtypes.SupplierServiceConfig{msg.ServiceConfig},
+	)
+
+	initialSupplierStake := cosmostypes.NewCoin(volatile.DenomuPOKT, math.ZeroInt())
+	foundSupplier, isFound := k.supplierKeeper.GetSupplier(ctx, shannonAccAddr.String())
+	if isFound {
+		initialSupplierStake = *foundSupplier.Stake
+	}
+
+	supplier, err := k.supplierKeeper.StakeSupplier(ctx, logger, msgStakeSupplier)
+	if err != nil {
+		// DEV_NOTE: StakeSupplier SHOULD ALWAYS return a gRPC status error.
+		return nil, err
+	}
+
+	// DEV_NOTE: When BOTH:
+	// - the claimed Shannon account is already staked as a supplier
+	// - the MsgClaimMorseSupplier stake amount ("default" or otherwise)
+	//   is less than the current supplier stake amount
+	// then, claimedSupplierStake is set to zero as it would otherwise result in a negative amount.
+	// This value is only used in event(s) and the msg response.
+	claimedSupplierStake, err := supplier.Stake.SafeSub(initialSupplierStake)
+	if err != nil {
+		if !strings.Contains(err.Error(), "negative coin amount") {
+			return nil, err
+		}
+		claimedSupplierStake = cosmostypes.NewInt64Coin(volatile.DenomuPOKT, 0)
+	}
+
+	claimedUnstakedTokens := morseClaimableAccount.TotalTokens().Sub(claimedSupplierStake)
+
+	// Emit an event which signals that the morse account has been claimed.
+	event := migrationtypes.EventMorseSupplierClaimed{
+		ShannonDestAddress:   msg.ShannonDestAddress,
+		MorseSrcAddress:      msg.MorseSrcAddress,
+		ServiceId:            supplier.GetServices()[0].GetServiceId(),
+		ClaimedBalance:       claimedUnstakedTokens,
+		ClaimedSupplierStake: claimedSupplierStake,
+		ClaimedAtHeight:      sdkCtx.BlockHeight(),
+		Supplier:             supplier,
+	}
+	if err = sdkCtx.EventManager().EmitTypedEvent(&event); err != nil {
+		return nil, status.Error(
+			codes.Internal,
+			migrationtypes.ErrMorseSupplierClaim.Wrapf(
+				"failed to emit event type %T: %v",
+				&event,
+				err,
+			).Error(),
+		)
+	}
+
+	// Return the response.
+	return &migrationtypes.MsgClaimMorseSupplierResponse{
+		MorseSrcAddress:      msg.MorseSrcAddress,
+		ServiceId:            supplier.Services[0].GetServiceId(),
+		ClaimedBalance:       claimedUnstakedTokens,
+		ClaimedSupplierStake: claimedSupplierStake,
+		ClaimedAtHeight:      sdkCtx.BlockHeight(),
```

Comment:
I'll re-review this whole function after we tend to the changes above. Might be easier to read from start to finish then.